quic/s2n-quic-core/src/dc/mod.rs

```diff
@@ -0,0 +1,146 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    connection::Limits,
+    crypto::tls::TlsSession,
+    event::{api::SocketAddress, IntoEvent as _},
+    inet,
+    path::MaxMtu,
+    stateless_reset,
+    transport::parameters::InitialFlowControlLimits,
+    varint::VarInt,
+};
+use core::time::Duration;
+
+pub use disabled::*;
+
+mod disabled;
+
+// dc versions supported by this code, in order of preference (SUPPORTED_VERSIONS[0] is most preferred)
+const SUPPORTED_VERSIONS: &[u32] = &[0x0];
+
+/// Called on the server to select the dc version to use (if any)
+///
+/// The server's version preference takes precedence
+pub fn select_version(client_supported_versions: &[u32]) -> Option<u32> {
+    SUPPORTED_VERSIONS
+        .iter()
+        .find(|&supported_version| client_supported_versions.contains(supported_version))
+        .copied()
+}
+
+/// The `dc::Endpoint` trait provides a way to support dc functionality
+pub trait Endpoint: 'static + Send {
+    /// If enabled, a dc version will attempt to be negotiated and dc-specific frames
+    /// will be processed. Otherwise, no dc version will be negotiated and dc-specific
+    /// frames received will result in a connection error.
+    const ENABLED: bool = true;
+
+    type Path: Path;
+
+    /// Called when a dc version has been negotiated for the given `ConnectionInfo`
+    fn new_path(&mut self, connection_info: &ConnectionInfo) -> Self::Path;
+}
+
+/// A dc path
+pub trait Path: 'static + Send {
+    /// Called when path secrets are ready to be derived from the given `TlsSession`
+    fn on_path_secrets_ready(&mut self, session: &impl TlsSession);
+
+    /// Called when a `DC_STATELESS_RESET_TOKENS` frame has been received from the peer
+    fn on_peer_stateless_reset_tokens<'a>(
+        &mut self,
+        stateless_reset_tokens: impl Iterator<Item = &'a stateless_reset::Token>,
+    );
+
+    /// Returns the stateless reset tokens to include in a `DC_STATELESS_RESET_TOKENS`
+    /// frame sent to the peer.
+    fn stateless_reset_tokens(&mut self) -> &[stateless_reset::Token];
```

Comment:
Trying to decide if this is going to be an issue if we're making the provider own the tokens (since they're being borrowed out of this function).

Reply:
works for me